---
title: "Class 17: Cancer Genomics Part 1"
author: "Serina Huang"
date: "`r format(Sys.time(), '%D')`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Section 1. The NCI Genomic Data Commons
The [Genomic Data Commons (GDC)](https://gdc.cancer.gov/about-gdc) website by the National Cancer Institute shares cancer genomic datasets.

Go on the [GDC Web Portal](https://portal.gdc.cancer.gov/) to answer the following questions:

Q1. How many _Cases_ (i.e. patient samples) have been found to have p53 mutations?

A1. 4,008.

Q2. What are the top 6 misssense mutations found in this gene? 
HINT: Scroll down to the _TP53 - Protein_ section and mouse over the displayed plot. For example, R175H is found in 156 cases.

A2. R175H, R273C, R248Q, R273H, R248W, R282W.

Q3. Which domain of the protein (as annotated by PFAM) do these mutations reside in?

A3. DNA-binding domain (pink bar).

Q4. What are the top 6 primary sites (i.e. cancer locations such as Lung, Brain, etc.) with p53 mutations and how many primary sites have p53 mutations been found in? HINT: Clicking on the number links in the Cancer Distribution section will take you to a summary of available data accross cases, genes, and mutations for p53. Looking at the _Cases_ data will give you a ranked listing of primary sites.

A4. Bronchus and lung; Ovary; Brain; Breast; Colon; Stomach. p53 mutations have been found is 44 primary sites.

Let's practice the same approach on a different protein: KRAS.

Q5. What is the most frequentely mutated position associated with cancer in the KRAS protein (i.e. the amino acid with the most mutations)?

A5. G12 (glycine). In other words, don't touch positions 12 and 13 in amino acid sequences!

Q6. Are KRas mutations common in Pancreatic Adenocarcinoma (i.e. is the Pancreas a common ‘primary site’ for KRas mutations?).

A6. Yes, 137 out of of 182 pancreatic cancer cases have Kras mutations. That's 75.27%!

Q7. What is the ‘TCGA project’ with the most KRas mutations?

A7. TCGA-PAAD.

Q7. What precent of cases for this ‘TCGA project’ have KRas mutations and what precent of cases have p53 mutations? HINT: Placing your mouse over the project bar in the Cancer Distribution panel will bring up a tooltip with useful summary data.

A7. 75.27% of TCGA-PAAD have KRas mutations, compared to 62.64% of TCGA-PAAD that have p53 mutations.

Q8. How many pancreatic adenocarcinoma cases (i.e. patients from the TCGA-PAAD project) have RNA-seq data available?

A8. 177 out of 185 TCGA-PAAD cases total.

Note: We can look at the mutation sites on the protein using VMD. When we do that, we see that the top few mutation sites are all close to the substructure that binds the GNP ligand.

### Section 2. The GenomicDataCommons R Package
The GDC Bioconductor package has functions that query. access, and mine the large genomic datasets on NCI-DGC, e.g. RNA-seq, WXS, SNP, etc. This is important because there are limitations to what we can deduct from the web portal.

We will also use the `maftools` package, which works on files in the Mutation Annotation Format (MAF).
```{r message = FALSE}
# One-time install of packages
# source("https://bioconductor.org/biocLite.R")
# biocLite(c("GenomicDataCommons", "maftools"))

# Load packages
library(GenomicDataCommons)
library(maftools)
```

In the next section, we will use `GenomciDataCommons` to identify somatic variants and `maftools` to do a high-level assessment of these variants.
```{r}
# Check on GDC status. It should say OK
GenomicDataCommons::status()
```

### Section 3. Querying the GDC from R
Let's narrow down the data we are interested in. This is called _querying GDC metadata_, where metadata refers to the description associated with the actual patient data (i.e. case) in the GDC.

There are 4 sets of metadata we can query with the `GenomicDataCommons` package: the `cases()`, `projects()`, `files()`, and `annotations()`. Let's start by using `cases()` to find the number of cases in each project.
```{r}
# This code snippet from the package will use %>% to pipe (transfer) the output of the LHS function into the input of the RHS function
# Piping is used to avoid nested functions, which can be hard to read or debug

# The cases() function, which takes no argument, is invoked
# The output of cases() is stored into a list called cases_by_project
# The cases_by_project object has 5 fields: fields, filters, facets, legacy, expand
# Next, the output cases_by_project is passed to the function facet() as its first argument
# facet() groups all cases by project ID
# According to the documentation of facet(), the second argument is a character vector in for the following aggregations() function
# Lastly, aggregations() counts up the cases in each project

cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)

# Access the data we want
View(cases_by_project$project.project_id)
```

Q9. Recreate the NCI barplot that displays the number of cases per project. The y-axis should be on a log scale and the x-axis should be rotated to allow vertical project labels.

A9.
```{r}
x <- cases_by_project$project.project_id
head(x)
  
# Test plot
barplot(x[,2], names.arg = x[,1], log = "y", las = 2)
# Bonus: What would col = c("lightblue", "red") do?
# Alternate blue and red within the length of the color vector

# Welps, that's ugly
# Make custom color vector and replot
colvec <- rep("lightblue", nrow(x))
paad.ind <- which(x == "TCGA-PAAD")
colvec[paad.ind] <- "red"

# Alternatively, this is probably better:
# (x$key == "TCGA-PAAD")
# Which one of these is TRUE?
# which(x$key == "TCGA-PAAD")

# Replot with sensible margins
par(mar = c(10, 5, 2, 1))
barplot(x[,2], names.arg = x[,1], log = "y", col = colvec, las = 2)
```

Great! Now, let's find all gene expression data files for all pancreatic cancer patients.
```{r}
# Another code snippet from the package vignette
file_records <- files() %>%
  filter(~ cases.project.project_id == "TCGA-PAAD" &
    data_type == "Gene Expression Quantification" &
    analysis.workflow_type == "HTSeq - Counts") %>%
response_all()

# This is essentially the same thing we can do on the web portal, but reproducible!
# View(file_records)
```

`file_records$results` contains a row for every RNA-seq data file from the TCGA-PAAD project. As of now, there are `{r}nrow(file_records$results)` projects.
```{r}
head(file_records$results)
```

We can download these R tools. For larger datasets like this one, the `transfer()` function can perform more robust data downloads by using the GDC transfer client (a separate command-line tool).

### Section 4. Variant Analysis
We can do a file query using Advanced Search on the NCI-GDC web portal to find all the MAF somation mutation files in the TCGA-PADD project. The search term is `cases.project.project_id in ["TCGA-PAAD"] and files.data_type in ["Masked Somatic Mutation"] and files.data_format in ["MAF"]`.

Q10. How many MAF files for the TCGA-PAAD project were found from this advanced web search?

A10. 4 files.

Alternatively, we can perform the same search in R with the `files()` function from the `GenomicDataCommons` package. Note: The last term focuses on only one of the MAF files for this project in GDC, namely the MuTect2 workflow variant calls.
```{r}
maf.files = files() %>%
    filter(~ cases.project.project_id == 'TCGA-PAAD' &
        data_type == 'Masked Somatic Mutation' &
        data_format == "MAF" &
        analysis.workflow_type == "MuTect2 Variant Aggregation and Masking") %>%
response_all()
```

```{r}
# View(maf.files)
attributes(maf.files)
head(maf.files$results)
```

Q11. What line in the above code could be modified to return all MAF files for the TCGA-PAAD project?

A11.
```{r}
allpaadmaf.files = files() %>%
    filter(~ cases.project.project_id == 'TCGA-PAAD' &
        data_format == "MAF") %>%
response_all()

# View(allpaadmaf.files)
```

Use the `ids()` function to pull out the unique identifier for our MAF file.
```{r}
uid <- ids(maf.files)
uid

# Download file to cache
maffile = gdcdata(uid)
maffile
```

This is the file we want to download. Do so with the `gdcdata()` function.
```{r}
# Download maf file
maffile = gdcdata(uid)
maffile
```

Now the maf file has been downloaded into our cache directory, we are ready for analysis using the `maftools` package.

#### MAF Analysis
```{r}
vars = read.maf(maf = maffile, verbose = FALSE)
```

Here are a few things `maftools` can do.

#### Plotting MAF Summary
The `plotmacSummary()` function makes a summary plot. For example, the __Variant Classification__ boxplot displays the variant types. The __Variants per sample__ stacked barplot displays the number of variants in each sample. We can modify the arguments of this function to display the average/median.
```{r}
plotmafSummary(maf =vars, rmOutlier = TRUE,
               addStat = 'median', dashboard = TRUE,
               titvRaw = FALSE)
```

#### Drawing Oncoplots
A useful summary representation of this data can be obtained from `oncoplot`, which is also called a waterfall plot.
```{r message = FALSE}
oncoplot(maf = vars, top = 10)
# Since the oncoplot is a large image, send this plot to a PNG or PDF.
# Send the oncoplot for our top 10 most frequently mutated genes to a PDF
pdf("oncogene_panc.pdf")
oncoplot(maf = vars, top = 10, fontSize = 12)
dev.off()
```

#### Oncostrip
`oncostrip` lets us visualize any set of genes by drawing the mutations of specific genes in each sample.
```{r}
oncostrip(maf = vars, genes = c("KRAS", "TP53"))
```

#### Lollipop Plot
```{r}
lollipopPlot(maf = vars, gene = "KRAS", showMutationRate = TRUE, domainLabelSize = 3)

lollipopPlot(maf = vars, gene = "TP53", showMutationRate = TRUE, domainLabelSize = 3)
```

------

### Summary
In part 1, we explored the Genomic Data Commons hosted by the National Cancer Institute in both its web portal and Bioconductor package forms. For the latter, there are multiple functions that can be used in conjunction with the `maftools` package to make informative plots.
