---
title: "Class 17: Cancer Genomics Part 2"
author: "Serina Huang"
date: "`r format(Sys.time(), '%D')`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Section 1. Protein Sequences from Healthy and Tumor Tissues
The following sequences are the results of an NGS analysis of a healthy vs. tumor tissue in one patient.

Q1. Identify sequence regions that contain all 9-mer peptides that are only found in the tumor.

A1. First, find the mutation sites by alignment (either by the web-based MUSCLE or the R `bio3d` package).
```{r}
library(bio3d)
p53seq <- read.fasta("https://bioboot.github.io/bimm143_F18//class-material/lecture18_sequences.fa")
p53seq
```

We want to find a 9-mer sequence that fits into the pocket in the MLC molecule; and within that 9-mer, there must be a mutation site that's specific to the tumor.

Each asterisk denotes a match. We see that the end of the mutant sequence is truncated. We can also eyeball a few other mutation sites, but this approach is error-prone and unproductive. It's also problematic if there are more than one mutation site and/or more than one patient.

Computationally, we can extract all possible 9-mer sequences containing a mutation site. In other words, the mutation can either be the first or the last of the 9-mer.

The `conserv()` function in `bio3d` that searches for conserved sequences. What we want, however, is the "non-conserved" sequences (of the tumor).
```{r}
score <- conserv(p53seq, method = "identity")
# Find the scores that are less than 1 and use which to get their indices
mut.ind <- which( (score < 1) )
# Verify the indices from the original fasta file
p53seq$ali[,1]
p53seq$ali[,2]
p53seq$ali[,3]
p53seq$ali[,41]

p53seq$ali[,mut.ind]
```

We just want the first four tumor-specific variants because we can't present "gaps" to the MHC molecule. We can take -8 and +8 positions of the mutant to get the series of 9-mers.

However, to do so, we need to get rid of the gaps first. Thankfully, there's already a `bio3d` function for it.
```{r}
gaps <- gap.inspect(p53seq)
# Indices for true gaps
notgaps.ind <- gaps$f.inds

# Now we have mut.ind and notgaps.ind
# Intersect the mutant sites vs. non-gaps
# The intersect function is %in%, i.e. in which of the index is mut.ind in notgaps.ind?
# Try in console: c(1:5) %in% c(4:10)
# Does dimension need to equal? Nope!
mut.ind_notgaps <- mut.ind[mut.ind %in% notgaps.ind]
mut.ind_notgaps
```

Now we can get the 8 before and 8 after positions for the 9-mers.
```{r}
start.position <- mut.ind_notgaps - 8
end.position <- mut.ind_notgaps + 8

start.position
end.position
```

Let's try to get our first 9-mer.
```{r}
p53seq$ali[ ,start.position[1]:end.position[1] ]
```

Aha! We confirm that at position 9, there's one of our mutation sites.
